homeassistant/components/adaptive_lighting/switch.py

```diff
@@ -0,0 +1,932 @@
+        # Tracks 'on' → 'off' state changes
+        self._off_to_on_event: Dict[str, Event] = {}
+        # Locks that prevent light adjusting when waiting for a light to 'turn_off'
+        self._locks: Dict[str, asyncio.Lock] = {}
+        # To identify that this integration made a change
+        self.__context = Context()  # self._context will be overwritten
+
+        # Set in self._update_attrs_and_maybe_adapt_lights
+        self._light_settings = {}
+
+        # Set and unset tracker in async_turn_on and async_turn_off
+        self.remove_listeners = []
+        _LOGGER.debug(
+            "%s: Setting up with '%s',"
+            " config_entry.data: '%s',"
+            " config_entry.options: '%s', converted to '%s'.",
+            self._name,
+            self._lights,
+            config_entry.data,
+            config_entry.options,
+            data,
+        )
+
+    @property
+    def entity_id(self):
+        """Return the entity ID of the switch."""
+        return self._entity_id
+
+    @property
+    def name(self):
+        """Return the name of the device if any."""
+        return f"Adaptive Lighting: {self._name}"
+
+    # @property
+    # def unique_id(self):
+    #     """Return the unique ID of entity."""
+    #     return self._name
+
+    @property
+    def is_on(self) -> Optional[bool]:
+        """Return true if adaptive lighting is on."""
+        return self._state
+
+    async def async_added_to_hass(self) -> None:
+        """Call when entity about to be added to hass."""
+        if self.hass.is_running:
+            await self._setup_listeners()
+        else:
+            self.hass.bus.async_listen_once(
+                EVENT_HOMEASSISTANT_START, self._setup_listeners
+            )
+        last_state = await self.async_get_last_state()
+        is_new_entry = last_state is None  # newly added to HA
+        if is_new_entry or last_state.state == STATE_ON:
+            await self.async_turn_on(adapt_lights=not self._only_once)
+        else:
+            self._state = False
+            assert not self.remove_listeners
+
+    def _expand_light_groups(self) -> None:
+        all_lights = _expand_light_groups(self.hass, self._lights)
+        self.turn_on_off_listener.lights.update(all_lights)
+        self._lights = list(all_lights)
+
+    async def _setup_listeners(self, _=None) -> None:
+        _LOGGER.debug("%s: Called '_setup_listeners'", self._name)
+        if not self.is_on or not self.hass.is_running:
+            _LOGGER.debug("%s: Cancelled '_setup_listeners'", self._name)
+            return
+
+        assert not self.remove_listeners
+
+        remove_interval = async_track_time_interval(
+            self.hass, self._async_update_at_interval, self._interval
+        )
+        remove_sleep = async_track_state_change_event(
+            self.hass,
+            self.sleep_mode_switch.entity_id,
+            self._sleep_state_event,
+        )
```

Comment:
If I comment out the about lines, the entity is properly removed when unloading. Otherwise, when changing options, I get this in my log messages:
```
2020-10-03 06:46:21 ERROR (MainThread) [homeassistant.components.switch] Entity id already exists - ignoring: switch.adaptive_lighting_wqeqw
```
Or when I manually delete the integration, the `switch.adaptive_lighting_wqeqw` remains until I restart HA.

Anyone any ideas why this happens?

Reply:
You must call that remove function when entity is removed. If you never touch that listener again, just pass it into self.async_on_remove()

Reply:
@elupus, thanks, however I am already doing that, see
* https://github.com/home-assistant/core/pull/40626/files/fc9d5f20946735c001d909ca7958ea1b897e9578#diff-b37e0ebaa19aa043e16de25d9ff440caR356
* https://github.com/home-assistant/core/pull/40626/files/fc9d5f20946735c001d909ca7958ea1b897e9578#diff-39fa7565115cb54d2880e4b9372130f5R102

I remove the listeners when turning off the switch.

Reply:
You should not access the entity from the entry unload. Your entity have its own unload callback.

Reply:
OK great. I added `async_will_remove_from_hass` and that fixed it.

I don't understand why though, @elupus, do you know why that worked?

Reply:
Not sure to be honest. Some lingering reference blocking the unload. But seems unlikely.